### ISSUE TITLE
sub-collections are being unhidden after export

### ISSUE THREAD

**Issue by dt192** [General-101/Halo-Asset-Blender-Development-Toolset#69]
Sub-collections are being unhidden after export regardless of having the export hidden options turned off.

e.g. Collection 2 here would be unhidden after export.
![2022-09-04_1662297133_blender](https://user-images.githubusercontent.com/2612048/188328710-e439a97e-ed9c-4ed6-827a-2b13c3a8beb6.png)

**Comment by General-101** [comment]
@jadshep This seems to be caused by gather_collection_resource function again. I've made some more changes that I believe fix the issue that I would like you to review.
```python
def gather_collection_resources(layer_collection, layer_collection_set, object_set, hidden_geo, nonrender_geo, include_this_collection = False):
    """
        Recursively gathers the relevant collections and objects for export, starting from a root layer collection.
        "include_this_collection" should be false for the root collection, else restoring visibility will cause everything to become visible regardless of stored value.
    """

    # Add collection to set of all included collections
    if include_this_collection:
        layer_collection_set.add(layer_collection)

    # Don't include anything from collection if hidden in an undesired way
    if not hidden_geo and not layer_collection.is_visible:
        return

    if not nonrender_geo and layer_collection.collection.hide_render:
        return

    # Add all of collection's objects when not hidden in an undesired way
    for obj in layer_collection.collection.objects:
        if not hidden_geo and obj.hide_viewport:
            continue

        if not hidden_geo and obj.hide_get():
            continue

        if not nonrender_geo and obj.hide_render:
            continue

        object_set.add(obj)

    # Recursively gather resources for child collections
    for collection in layer_collection.children:
        gather_collection_resources(collection, layer_collection_set, object_set, hidden_geo, nonrender_geo, True)
```
I've removed the exclude check from the if statement since is_visible takes care of that and I've changed the order of operations so that layer_collections get added to the list before it returns for something.

**Comment by jadshep** [comment]
@General-101 It appears that setting `exclude` for a layer collection also sets it for all of its children, which I didn't account for when writing that section.

What's causing the issue is that the functions that store/restore visibility only act on resources included in the export, and since that excluded collection isn't being exported its visibility states aren't being stored/restored, so when the exclude state for its parent collection is restored Blender sets it for all of its children, including the one that's not being stored/restored.

I think the solution is to just store and restore the visibility attributes for all collections in the current layer, instead of just the ones being exported. I should be able to do that sometime today.

**Comment by jadshep** [comment]
Also, my thinking for filtering the output is:
- Filter out anything hidden from the viewport UNLESS hidden geo is set
- Filter out anything hidden from render UNLESS nonrender geo is set
- Filter out any excluded collections, no matter what

Figuring that "exclude" is pretty explicit. But maybe it needs yet another option on output like "Export excluded collections"?

**Comment by General-101** [comment]
The expectation from community members who I've talked to regarding hidden geometry is that the export hidden geometry option will put out everything. I'm not super against another option but I feel it starts to border on making exporting a bit complicated vs a simple yes and no.

**Comment by jadshep** [comment]
I do think "Export excluded collections" is unneeded and pointlessly complicated.

Has the addition of the non-render option been a problem for others? It's defaulted to true, which should have the hidden geo act the same as it used to.

**Comment by jadshep** [comment]
Also #70 should fix this issue with few changes.

It was setting exclude to false for export, and storing/restoring, even though any included collection wouldn't be excluded by definition.

**Comment by General-101** [comment]
Haven't heard any complaints personally. Thank you for the quick fix again. I'll give this a look in a bit.

**Comment by dt192** [comment]
imo there should be two types of hidden, one for things like portals and skybox that allows it to be hidden from viewport all the time but still imported and another type for stuff you want to not include.

**Comment by dt192** [comment]
Actually that doesn't seem to be working as i would expect.

In this case, I would expect collection 1 and 2 to not export as they are hidden/disabled and I would expect collection 3, skybox and portals to all export, but nothing exports. Reading the icon descriptions the eye icon and screen icon are aparently just local and global options of hide from viewport, so I'm not sure what is being considered non-render for the override above. If it's 'Disable in renders' then that isn't very useful as the portals/skybox etc are still in my way in the viewport.

![2022-09-05_1662336886_blender](https://user-images.githubusercontent.com/2612048/188339089-3d08163f-b1ad-44e0-8610-1b8332ccfc47.png)

**Comment by jadshep** [comment]
@dt192 I'm confused why everything is greyed out, indicating they're all hidden, even though they aren't... I don't know what the deal is there.

Regarding the "render" terminology, it isn't really applicable since we generally don't render anything out for map making. Since that toggle isn't useful as intended I just used it as an export filtering option that's disconnected from what's visible in the viewport.

My workflow is to have "Export Hidden Geometry" ON and "Export Non-render Geometry" OFF. That way I pick what's exported just with the render visibility and don't have to worry about what's visible in the viewport.

**Comment by General-101** [comment]
> Actually that doesn't seem to be working as i would expect.
> 
> In this case, I would expect collection 1 and 2 to not export as they are hidden/disabled and I would expect collection 3, skybox and portals to all export, but nothing exports. Reading the icon descriptions the eye icon and screen icon are aparently just local and global options of hide from viewport, so I'm not sure what is being considered non-render for the override above. If it's 'Disable in renders' then that isn't very useful as the portals/skybox etc are still in my way in the viewport.
> 
> ![2022-09-05_1662336886_blender](https://user-images.githubusercontent.com/2612048/188339089-3d08163f-b1ad-44e0-8610-1b8332ccfc47.png)

Like the open PR doesn't behave as you'd expect? What about the code snippet I posted at the start of this issue?

**Comment by dt192** [comment]
Ok. That seems like a good solution to me.

I think the issue is that the option label doesn't indicate what it actually does. 'Export Non-Render Geometry' indicates that it's an override that exports geometry that has render visibility disabled, however that was the default behaviour of the ass exporter (it exported geo regardless of render visibility toggle), which is why I couldn't see the use.

I think the option be better named 'Exclude Non-Render Geometry From Export', or similar, as this is the actual function...
A better way might be to have a drop down or toggle instead of 2 check boxes such as this...
![2022-09-05_1662367588_firefox](https://user-images.githubusercontent.com/2612048/188408268-274d7702-6d99-40b5-b40c-0544fb6b1f78.png)

**Comment by General-101** [comment]
Fixed in 3915783
